Sources/NIOWebSocket/WebSocketFrame.swift

```diff
@@ -0,0 +1,270 @@
+    ///
+    /// On frames received from the network, this data is not necessarily unmasked. This is to provide as much
+    /// information as possible. If unmasked data is desired, either use the computed `unmaskedExtensionData` property to
+    /// obtain it, or transform this data directly by calling `extensionData.unmask(maskKey)`.
+    public var extensionData: ByteBuffer? = nil
+
+    /// The unmasked application data.
+    ///
+    /// If a masking key is present on the frame, this property will automatically unmask the underlying data
+    /// and return the unmasked data to the user. This is a convenience method that should only be used when
+    /// persisting the underlying masked data is worthwhile: otherwise, performance will often be better to
+    /// manually unmask the data with `data.unmask(maskKey)`.
+    public var unmaskedData: ByteBuffer {
+        get {
+            guard let maskKey = self.maskKey else {
+                return self.data
+            }
+            var data = self.data
+            data.unmask(maskKey, indexOffset: (self.extensionData?.readableBytes ?? 0) % 4)
+            return data
+        }
+    }
+
+    /// The unmasked extension data.
+    ///
+    /// If a masking key is present on the frame, this property will automatically unmask the underlying data
+    /// and return the unmasked data to the user. This is a convenience method that should only be used when
+    /// persisting the underlying masked data is worthwhile: otherwise, performance will often be better to
+    /// manually unmask the data with `data.unmask(maskKey)`.
+    public var unmaskedExtensionData: ByteBuffer? {
+        get {
+            guard let maskKey = self.maskKey else {
+                return self.extensionData
+            }
+            var extensionData = self.extensionData
+            extensionData?.unmask(maskKey)
+            return extensionData
+        }
+    }
+
+    /// Creates an empty `WebSocketFrame`.
+    ///
+    /// - parameters:
+    ///     - allocator: The `ByteBufferAllocator` to use when editing the empty buffers.
+    public init(allocator: ByteBufferAllocator) {
+        self.data = allocator.buffer(capacity: 0)
+    }
+
+    /// Create a `WebSocketFrame` with the given properties.
+    ///
+    /// - parameters:
+    ///     - fin: The value of the `fin` bit. Defaults to `false`.
+    ///     - rsv1: The value of the first reserved bit. Defaults to `false`.
+    ///     - rsv2: The value of the second reserved bit. Defaults to `false`.
+    ///     - rsv3: The value of the third reserved bit. Defaults to `false`.
+    ///     - opcode: The opcode for the frame. Defaults to `.continuation`.
+    ///     - maskKey: The masking key for the frame, if any. Defaults to `nil`.
+    ///     - data: The application data for the frame.
+    ///     - extensionData: The extension data for the frame.
+    public init(fin: Bool = false, rsv1: Bool = false, rsv2: Bool = false, rsv3: Bool = false,
```

Comment:
doesn't this create like 256 methods? Linker might not be super happy :)

Reply:
My understanding is that it creates 9 methods: one for the `init`, and then 8 for the default values of each of the parameters.

Reply:
and each of the combinations? It's 8 parameters and all the different combinations need to be available, no? That I think would be 2^8 = 256, no?

Reply:
No, I don't think so. The output of disassembling the *unoptimized* `.o` file for this `.swift` and passing it through `swift-demangle | grep "NIOWebSocket.WebSocketFrame.init"` gives us this:

```
_NIOWebSocket.WebSocketFrame.init(allocator: NIO.ByteBufferAllocator) -> NIOWebSocket.WebSocketFrame:
_default argument 0 of NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_default argument 1 of NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_default argument 2 of NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_default argument 3 of NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_default argument 4 of NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_default argument 5 of NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_default argument 7 of NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_NIOWebSocket.WebSocketFrame.init(fin: Swift.Bool, rsv1: Swift.Bool, rsv2: Swift.Bool, rsv3: Swift.Bool, opcode: NIOWebSocket.WebSocketOpcode, maskKey: NIOWebSocket.WebSocketMaskingKey?, data: NIO.ByteBuffer, extensionData: NIO.ByteBuffer?) -> NIOWebSocket.WebSocketFrame:
_NIOWebSocket.WebSocketFrame.init(firstByte: Swift.UInt8, maskKey: NIOWebSocket.WebSocketMaskingKey?, applicationData: NIO.ByteBuffer) -> NIOWebSocket.WebSocketFrame:
```

Ignoring the first and last element, which are for the other two constructors, you can see that there is one symbol for the function, and one for each default argument.

The compiler doesn't synthesise one symbol per possible combination of default arguments, that would make them exceedingly problematic. Instead it does the *slightly* less problematic thing of serialising only one symbol for the function that requires all the arguments, and then providing symbols that allow the caller to obtain the default value. Thus, if we have this:

```swift
func testFunc(x: Int, y: Int = 0, z: Int = 0) { }
testFunc(x: 5)
```

The invocation of `testFunc` will actually generate three function calls in the underlying unoptimised assembly:

```
y = getDefaultArgumentForY()
z = getDefaultArgumentForZ()
testFunc(x: 5, y: y, z: z)
```

I believe the Swift team have enhanced Swift 4.1 to always-inline these argument generators. See [this forum thread](https://forums.swift.org/t/idea-question-default-arguments-vs-abi-vs-code-size/10823) for more.

Reply:
thanks! that's perfect and given and also 4.1 is almost out I think :)